KAFKA-16753: Implement share acknowledge API in partition (KIP-932)

apoorvmittal10 opened the pull request:
The share-partition leader keeps track of the state and delivery attempts for in-flight records. However, delivery attempts tracking follows atleast-once semantics. 

The consumer processes the records and acknowledges them upon successful consumption. This successful attempt triggers a transition to the "Acknowledged" state.

The code implements the functionality to acknowledge the offset/batches in the request to in-memory cached data.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-16753: Implement share acknowledge API in partition (KIP-932)

apoorvmittal10 commented:
@AndrewJSchofield The PR build will fail untill we have APIs defined for Persister interface, I ll update the code once dependent PR is merged.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-16753

apoorvmittal10 committed:
Integrating share persister

apoorvmittal10 commented:
The PR is ready for review, I have rebased it to merged Persister PR.

apoorvmittal10 committed:
Refactored persister code

apoorvmittal10 committed:
Fixing checkstyle error

apoorvmittal10 commented:
@omkreddy Build completed with unrelated tests failure.

omkreddy reviewed (approved):
@apoorvmittal10 Thanks for the PR. LGTM

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+            }
+
+            NavigableMap<Long, InFlightBatch> subMap = cachedState.subMap(floorOffset.getKey(), true, batch.lastOffset(), true);
+            // Validate if the request batch has the first offset less than the last offset of the last
```
Should "less than" be "greater than"?

apoorvmittal10 replied:
Thanks for the catch, corrected the comment.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+                return RecordState.AVAILABLE;
+            case 3 /* REJECT */:
+            case 0 /* GAP */:
+                return RecordState.ARCHIVED;
```
Hmm, 0 means the state of the record hasn't changed. Why does it translate to `RecordState.ARCHIVED`?

apoorvmittal10 replied:
0 means that a gap has been identified by the client, i.e. offset not present. The in-memory state for the gap offsets directly marks the offset as archived so it won't be processed again.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+                    .setTopicsData(Collections.singletonList(new TopicData<>(topicIdPartition.topicId(),
+                        Collections.singletonList(PartitionFactory.newPartitionStateBatchData(
+                            topicIdPartition.partition(), stateEpoch, startOffset, 0, stateBatches))))
+                    ).build()).build()).get();
```
This blocks the request thread, which is not ideal.

apoorvmittal10 replied:
I agree this is not ideal. The load of share partition with persister read is temporary and rather the code will be moved to Share Coordinator which will load the share partitions in near future.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+            // Validate if the request batch has the last offset greater than the last offset of
+            // the last fetched cached batch, then there will be offsets in the request than cannot
+            // be found in the fetched batches.
+            if (isLastBatch && batch.lastOffset() > subMap.lastEntry().getValue().lastOffset) {
```
Hmm, why is the check only applied on the last batch instead of every batch?

apoorvmittal10 replied:
Hmmm, I missed it. You are right, this could happen. I have corrected it and added a test as well.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+        }
+    }
+
+    private Optional<Throwable> acknowledgementBatchRecords(
```
acknowledgementBatchRecords => acknowledgeBatchRecords ?

apoorvmittal10 replied:
Thanks, done.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+        }
+        RecordState startOffsetState = entry.getValue().offsetState == null ?
+            entry.getValue().batchState() :
+            entry.getValue().offsetState().get(startOffset).state();
```
In a compacted topic, the baseOffset may not be the same as the firstOffset in a batch. startOffset is derived from baseOffset while offsetState() is initialized based on firstOffset. So, it seems that we could hit a NullPointerException here.

apoorvmittal10 replied:
For compacted topics there could be 2 possibilities:
1) The gap exists within the batch i.e. batch fetch from 10 - 20 and gap exists at 15 offset
2) The gap exists outside of the fetched batches i.e. fetched batches are 10-20, 23-35, the gap exists at 21 and 22.

For 1 the gap is identified by the client where offset is moved to Archived state and for 2 the gaps are never filled up in in-memory skip list.

However as the state is fetched directly on startOffset hence there might be a case where we can have NPE. I have created a task for myself to work on it: https://issues.apache.org/jira/browse/KAFKA-17351

junrao replied:
Just to be clear. The case that I was referring to is the gap between baseOffset and firstOffset. Consider a batch with baseOffset 100 and firstOffset of 105. In the following code, InFlightBatch.offsetState will start at offset 105.

```
        private void maybeInitializeOffsetStateUpdate() {
            if (offsetState == null) {
                offsetState = new ConcurrentSkipListMap<>();
                // The offset state map is not initialized hence initialize the state of the offsets
                // from the first offset to the last offset. Mark the batch inflightState to null as
                // the state of the records is maintained in the offset state map now.
                for (long offset = this.firstOffset; offset <= this.lastOffset; offset++) {
``` 

We set startOffset from either baseOffset of a batch or lastOffset of a batch + 1. So startOffset could be 100 or less than 100 (if previous batch has more than one offset gap at the end). In either case, we will hit `NullPointerException` when calling `offsetState().get(startOffset)`.

apoorvmittal10 replied:
I might be mistaken but still trying to understand how this could happen. The base offset will always be the first offset of the batch, the situation where baseOffset and firstOffset within a batch should not exist.

junrao replied:
This is the tricky part with compacted topics. The following is the code in `MemoryRecords` for generating a new batch for retained records after cleaning. The code uses the baseOffset from the original batch as the baseOffset, which can be different from the offset of the first retained record.

```
    private static MemoryRecordsBuilder buildRetainedRecordsInto(RecordBatch originalBatch,
                                                                 List<Record> retainedRecords,
                                                                 ByteBufferOutputStream bufferOutputStream,
                                                                 final long deleteHorizonMs) {
        byte magic = originalBatch.magic();
        Compression compression = Compression.of(originalBatch.compressionType()).build();
        TimestampType timestampType = originalBatch.timestampType();
        long logAppendTime = timestampType == TimestampType.LOG_APPEND_TIME ?
                originalBatch.maxTimestamp() : RecordBatch.NO_TIMESTAMP;
        long baseOffset = magic >= RecordBatch.MAGIC_VALUE_V2 ?
                originalBatch.baseOffset() : retainedRecords.get(0).offset();

        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferOutputStream, magic,
                compression, timestampType, baseOffset, logAppendTime, originalBatch.producerId(),
                originalBatch.producerEpoch(), originalBatch.baseSequence(), originalBatch.isTransactional(),
                originalBatch.isControlBatch(), originalBatch.partitionLeaderEpoch(), bufferOutputStream.limit(), deleteHorizonMs);

        for (Record record : retainedRecords)
            builder.append(record);

```

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Left a few late review comments.

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the PR. Left a few late review comments.

@junrao Thanks for the review. I have addressed the comments here: https://github.com/apache/kafka/pull/16894
